basic validator implementation

wesleytruong opened the pull request:
Update PR Summary:
Implements a validator that can be easily plugged into the training loop and configured from the job specific config file.
Changes:
- Created validation section in job_config with enabled, dataset, freq, and steps fields
- Created a builder function for validator in train_spec
- Created a separate builder function for validation dataset in hf_dataset.py
- Created validator class
  - Validator class initializes a build_validation_hf_loader but leaves this dataloader function unexposed to the train_spec
  - Validator class supports ddp, fsdp, cp, and tp (but not pp yet)
- Integrated validation call into training loop
- Creates an integration test to test parallelization

Updated tests training the same base model weights from a seed checkpoint:

| FSDP=2 | FSDP=2,TP=4 |
| --- | --- |
| <img width="978" alt="Screenshot 2025-07-09 at 4 33 53 PM" src="https://github.com/user-attachments/assets/a1fa9fa7-df2f-4302-aa4a-d556a5699ba9" /> | <img width="978" alt="Screenshot 2025-07-09 at 4 33 53 PM" src="https://github.com/user-attachments/assets/a1fa9fa7-df2f-4302-aa4a-d556a5699ba9" /> |

| FSDP=2,CP=4 | FSDP=2,TP=2,CP=2 |
| --- | --- |
| <img width="972" alt="Screenshot 2025-07-09 at 4 39 35 PM" src="https://github.com/user-attachments/assets/56d62841-5841-4969-85b1-803705892465" /> | <img width="970" alt="Screenshot 2025-07-09 at 4 28 57 PM" src="https://github.com/user-attachments/assets/f7d33fa8-ca2c-48f1-931c-8d4c017a47ce" /> |

wwwjn commented on torchtitan/config_manager.py:
```diff
@@ -657,6 +657,24 @@ class Experimental:
 
+@dataclass
+class Validation:
+    enabled: bool = False
```
Do we need to have a parameter (eg, eval_frequence, integer) to control how many training steps we want to run eval?

tianyu-l commented on torchtitan/components/validate.py:
```diff
@@ -0,0 +1,121 @@
+            logger.warning("No validation batches processed")
+
+        # Set model back to train mode
+        self.model.train()
```
let's put this as the last line of this method

wesleytruong commented:
I've cleaned up the code according to your comments and added support for the validation frequency and steps. I also left streaming=True in the c4_validation dataset since otherwise it downloads the entire training dataset too. @tianyu-l

tianyu-l commented on torchtitan/components/validate.py:
```diff
@@ -0,0 +1,131 @@
+        num_val_steps = 0
+
+        with torch.no_grad():
+            try:
```
I believe you don't need this try-catch because StopIteration will be automatically captured by for loop safely.

runame commented on torchtitan/components/validate.py:
```diff
@@ -0,0 +1,131 @@
+        model_parts: list[nn.Module],
+    ) -> dict[str, float]:
+        # Set model to eval mode
+        model = model_parts[0]
```
Is there a reason to not support all parallelisms besides PP here?

runame commented on torchtitan/components/validate.py:
```diff
@@ -0,0 +1,131 @@
+            except StopIteration:
+                logger.info("Validation dataloader exhausted")
+
+        # Compute average loss
```
This assumes that the number of tokens is the same for every batch. Maybe either manually keep track of the total number of tokens or at least add a NOTE that highlights this assumption.

runame commented on torchtitan/components/validate.py:
```diff
@@ -0,0 +1,131 @@
+        # Compute average loss
+        if num_batches > 0:
+            average_loss = total_loss / num_batches
+        else:
```
I think this code path should never be used, you could guarantee this (ignoring the case of an empty dataloader) by adding a `__post_init__` to the `Validation` dataclass that verifies that all values are valid, e.g., `val_steps > 0`.

runame commented on torchtitan/components/validate.py:
```diff
@@ -0,0 +1,131 @@
+        # Set model back to train mode
+        model.train()
+
+        return {"validation_loss": average_loss}
```
The `average_loss` is the local loss for each rank, but should still be all-reduced across ranks.

runame replied:
Could you change this to `"validation/loss"`? This is important for how wandb represents the metrics and allows you to add more metrics to the same section via `"validation/<you-new-metric>"` later on.

runame commented on torchtitan/components/validate.py:
```diff
@@ -0,0 +1,131 @@
+                    total_loss += loss.item()
+                    num_batches += 1
+
+                    num_val_steps += 1
```
Is there a reason you use separate counters for `num_batches` and `num_val_steps`? Also, you could use this instead:
```python
for step, (input_dict, labels) in enumerate(self.validation_dataloader):
```
Here, `step` replaces `num_batches` and `num_val_steps`. You would also have to change `num_val_steps >= self.job_config.validation.val_steps` to `step > self.job_config.validation.val_steps` above.

runame commented on torchtitan/components/validate.py:
```diff
@@ -0,0 +1,131 @@
+        device_type = utils.device_type
+        num_val_steps = 0
+
+        with torch.no_grad():
```
Nit: you can also use this as a decorator instead, so you don't have to indent your code as much.
```python
@torch.no_grad()
def validate(
```

runame reviewed (commented):
Thanks for implementing this, this will be very useful!

You can take a look at [these changes](https://github.com/pytorch/torchtitan/compare/main...runame:torchtitan:validation) for some inspiration for addressing some of my comments.

tianyu-l commented on torchtitan/config_manager.py:
```diff
@@ -657,6 +657,30 @@ class Experimental:
 
+@dataclass
+class Validation:
+    enabled: bool = False
```
I think we can keep this `enabled` to be consistent with other configs in torchtitan -- this sounds more like a style thing?

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -193,3 +200,34 @@ def build_hf_dataloader(
     )
+
+
+def build_hf_validation_dataloader(
```
The reason we probably don't want to change this interface is:
people plug in their own data loader, and they want it to be general enough https://github.com/pytorch/torchtitan/blob/main/torchtitan/train.py#L133-L138
We used to have something closer to what you proposed, but changed due to their requests.

I think it's ok we make this compromise for that purpose.

fegin commented on torchtitan/components/validate.py:
```diff
@@ -0,0 +1,131 @@
+        model_parts: list[nn.Module],
+    ) -> dict[str, float]:
+        # Set model to eval mode
+        model = model_parts[0]
```
Does this PR assume the model is not sharded and will handle model being sharded later? If so can we raise an exception if dp_shard > 1? If we do support FSDP, then you will need to be careful about `reshard_after_forward` value as ensure the parameters are sharded before leaving `validate()`. Otherwise, checkpointing will be broken.

fegin commented on torchtitan/train.py:
```diff
@@ -463,6 +477,12 @@ def train_step(
         else:
             global_avg_loss = global_max_loss = loss.detach().item()
 
+        # Run validation if validator is available
```
As my comment above, please check if validation breaks checkpointing if FSDP is used.

wesleytruong commented:
To show the validation parallelism is correct I followed https://github.com/pytorch/torchtitan/blob/main/docs/debugging.md#reproducibility-between-runs, to compare training the same model with different parallelisms and show they have the same loss
FSDP=2:
<img width="966" alt="Screenshot 2025-07-08 at 3 28 17 PM" src="https://github.com/user-attachments/assets/ab753c21-1e41-43b9-8df7-f5110fe95017" />
FSDP=2, TP=2:
<img width="970" alt="Screenshot 2025-07-08 at 3 30 17 PM" src="https://github.com/user-attachments/assets/c3e8a05a-2ed6-4d8c-9a0c-10e6622ecac8" />
FSDP=2, TP=2, CP=2:
<img width="971" alt="Screenshot 2025-07-08 at 3 34 06 PM" src="https://github.com/user-attachments/assets/baa4491c-dfa7-4a78-a397-dc066d3bdf00" />

tianyu-l reviewed (commented):
Looks good, could you edit the PR summary with latest test results (FSDP + CP + TP gives the same results as FSDP / 1GPU results)
